pkg/policy/cell/identity_updater.go

```diff
@@ -0,0 +1,223 @@
+
+// identityAllocatorOwner is used to break the circular dependency between
+// CachingIdentityAllocator and policy.Repository.
+type identityUpdater struct {
+	logger    *slog.Logger
+	policy    policy.PolicyRepository
+	epmanager endpointmanager.EndpointManager
+
+	identityHandlers []identity.UpdateIdentities
+
+	// set of notification waitgroups to wait in for batched UpdatePolicyMaps,
+	// and a mutex to protect for writing
+	wgsLock        lock.Mutex
+	wgs            []*sync.WaitGroup
+	firstStartTime time.Time // the start time for the first batched update
+
+	updatePolicyMaps job.Trigger
+}
+
+type identityAllocatorParams struct {
+	cell.In
+
+	Log              *slog.Logger
+	Registry         job.Registry
+	Lifecycle        cell.Lifecycle
+	PolicyRepository policy.PolicyRepository
+	EPManager        endpointmanager.EndpointManager
+	Health           cell.Health
+	Metrics          *updaterMetrics
+
+	IdentityHandlers []identity.UpdateIdentities `group:"identity-handlers"`
+}
+
+func newIdentityUpdater(params identityAllocatorParams) IdentityUpdater {
+	i := &identityUpdater{
+		logger:    params.Log,
+		policy:    params.PolicyRepository,
+		epmanager: params.EPManager,
+
+		identityHandlers: params.IdentityHandlers,
+	}
+	i.updatePolicyMaps = job.NewTrigger()
+	jg := params.Registry.NewGroup(params.Health, params.Lifecycle, job.WithMetrics(params.Metrics), job.WithLogger(params.Log))
+	jg.Add(job.Timer("id-alloc-update-policy-maps", i.doUpdatePolicyMaps,
+		/* no interval, only on trigger */ 0, job.WithTrigger(i.updatePolicyMaps)))
+
+	return i
+}
+
+// UpdateIdentities informs the policy package of all identity changes
+// and also triggers policy updates. Updates to the endpoints are batched.
+//
+// The caller is responsible for making sure the same identity is not
+// present in both 'added' and 'deleted'.
+func (i *identityUpdater) UpdateIdentities(added, deleted identity.IdentityMap) {
+	// Have we already seen this exact set of updates? If so, we can skip.
+	// This happens when a global identity is allocated locally (for an Endpoint).
+	// We will add the identity twice; once directly from the endpoint creation,
+	// and again from the global identity watcher (k8s or kvstore).
+	if i.policy.GetSelectorCache().CanSkipUpdate(added, deleted) {
+		i.logger.Debug("Skipping no-op identity update")
+		return
+	}
+
+	start := time.Now()
+
+	i.logger.Info(
+		"Processing identity update",
+		logfields.AddedPolicyID, slices.Collect(maps.Keys(added)),
+		logfields.DeletedPolicyID, slices.Collect(maps.Keys(deleted)),
+	)
+
+	wg := &sync.WaitGroup{}
+	for _, handler := range i.identityHandlers {
+		handler.UpdateIdentities(added, deleted, wg)
+	}
+	// Invoke policy selector cache always as the last handler
+	// This synchronously updates the SelectorCache and queues an incremental
+	// update to any selectors. The waitgroup is closed when all endpoints
+	// have been notified.
+	i.policy.GetSelectorCache().UpdateIdentities(added, deleted, wg)
```

Comment:
This undoes the `mutated` check introduced in the previous commit. This is not intentional, right?

Edit: Ah, I see the next commit re-introduces it. The commit message here is a bit misleading (`The code
is otherwise unchanged.`), but ultimately I think it's fine.

Reply:
Oh whoops, that's awkward. I'll clean up the commit history.